client/src/main/scala/org/http4s/client/middleware/CookieJar.scala

```diff
@@ -0,0 +1,205 @@
+
+  /**
+    * Default Expiration Approach, Removes Expired Cookies
+    */
+  def evictExpired: F[Unit]
+
+  /**
+    * Available for Use To Relieve Memory Pressure
+    */
+  def evictAll: F[Unit]
+
+  /**
+    * Add Cookie to the cookie jar
+    */
+  def addCookie(c: ResponseCookie, uri: Uri): F[Unit]
+
+  /**
+    * Enrich a Request with the cookies available
+    */
+  def enrichRequest[G[_]](r: Request[G]): F[Request[G]]
+
+}
+
+/**
+  * Cookie Jar Companion Object
+  * Contains constructors for client middleware or raw
+  * jar creation, as well as the middleware
+ **/
+object CookieJar {
+
+  /**
+    * Middleware Constructor Using a Provided [[CookieJar]].
+   **/
+  def apply[F[_]: Sync](
+      alg: CookieJar[F]
+  )(
+      client: Client[F]
+  ): Client[F] =
+    Client { req =>
+      for {
+        _ <- Resource.liftF(alg.evictExpired)
+        modRequest <- Resource.liftF(alg.enrichRequest(req))
+        out <- client.run(modRequest)
+        _ <- Resource.liftF(
+          out.cookies
+            .map(r => r.domain.fold(r.copy(domain = req.uri.host.map(_.value)))(_ => r))
+            .traverse_(alg.addCookie(_, req.uri))
+        )
+      } yield out
+    }
+
+  /**
+    * Constructor which builds a non-exposed CookieJar
+    * and applies it to the client.
+   **/
+  def impl[F[_]: Sync: Timer](c: Client[F]): F[Client[F]] =
+    in[F, F](c)
+
+  /**
+    * Like [[impl]] except it allows the creation of the middleware in a
+    * different HKT than the client is in.
+   **/
+  def in[F[_]: Sync: Timer, G[_]: Sync](c: Client[F]): G[Client[F]] =
+    jarIn[F, G].map(apply(_)(c))
+
+  /**
+    * Jar Constructor
+   **/
+  def jarImpl[F[_]: Sync: Clock]: F[CookieJar[F]] =
+    jarIn[F, F]
+
+  /**
+    * Like [[jarImpl]] except it allows the creation of the CookieJar in a
+    * different HKT than the client is in.
+   **/
+  def jarIn[F[_]: Sync: Clock, G[_]: Sync]: G[CookieJar[F]] =
+    Ref.in[G, F, Map[CookieKey, CookieValue]](Map.empty).map { ref =>
+      new CookieJarRefImpl[F](ref) {}
+    }
+
+  private[CookieJar] class CookieJarRefImpl[F[_]: Sync: Clock](
+      ref: Ref[F, Map[CookieKey, CookieValue]]
+  ) extends CookieJar[F] {
+    override def evictExpired: F[Unit] =
+      for {
+        now <- currentHttpDate
+        out <- ref.update(
+          _.filter { t =>
+            now <= t._2.expiresAt
+          }
+        )
+      } yield out
+
+    override def evictAll: F[Unit] = ref.set(Map.empty)
+
+    override def addCookie(c: ResponseCookie, uri: Uri): F[Unit] =
+      for {
+        now <- currentHttpDate
+        out <- ref.update(extractFromResponseCookie(_)(c, now, uri))
+      } yield out
+
+    override def enrichRequest[N[_]](r: Request[N]): F[Request[N]] =
+      for {
+        cookies <- ref.get.map(_.map(_._2.cookie).toList)
+      } yield cookiesForRequest(r, cookies)
+        .foldLeft(r) { case (req, cookie) => req.addCookie(cookie) }
+  }
+
+  private[middleware] final case class CookieKey(
```

Comment:
Is `CookieJar` not tight enough scope here?

Reply:
Was future proofing for eventual testing I'm sure might become necessary.